### PR TITLE
[Frontend][Caffe] Introduce Caffe frontend to TVM

### PR DESCRIPTION
**Pull request by fernchen**
### Background & Motivation

[Caffe](<https://caffe.berkeleyvision.org/>) is a deep learning framework made with expression, speed, and modularity in mind. Because of its simplicity, good scalability, fast and other characteristics, it is favored by many people. According to [riselab]( https://rise.cs.berkeley.edu/) who makes statistics of papers collected in arxiv.org, Caffe is ranked in the top four in the deep learning framework, which shows to some extent that Caffe's user base is still large, please refer to [blog](<https://www.oreilly.com/content/one-simple-graphic-researchers-love-pytorch-and-tensorflow/>). In addition, according to our company's research on the market, the demand for Caffe in the production environment is still strong, and many models based on Caffe need to be deployed. For example, existing deployment frameworks, such as [MNN](<https://github.com/alibaba/MNN>), [NCNN](<https://github.com/Tencent/ncnn>), [MACE](<https://github.com/XiaoMi/mace>), etc., directly support the deployment of Caffe.

TVM only supports caffe2 at present, and the difference between Caffe and caffe2 is quite large. At present, there are two ways to deploy Caffe model in TVM: one is to convert Caffe model to Tensorflow or Pytorch model, the other is to convert Caffe model to onnx and then to relay IR. The two methods are essentially the same. They are all indirectly converted to relay IR through the third-party transformation. However, the problem is that some ops will fail in the process of model transformation, and even the result of transfer out may be different.

Based on the above situation, we decided to open our Caffe frontend codes, hoping to enrich the use scenarios of TVM.

### Implementation Approach

The whole process of Caffe front end importing model is divided into:

1. Read Model：The model graph and related parameters are read through the protobuffer API of Caffe;
2. Rebuild Graph：Traverse the graph, then replace the top of the in-place layer with the name of the layer, and update all related layers at the same time;
3. Model Conversion：Read the parameters of each layer and convert them into corresponding TVM OP and parameters;
4. Layer Fusion：fuse batchnorm and scale layers;
5. Convert to Relay IR：It mainly includes its module, params and the real name of the output layer。

Finally, we can import the Caffe model as follows:

```Python
from google.protobuf import text_format
from tvm.relay.frontend import caffe_pb2 as pb

init_net = pb.NetParameter()
predict_net = pb.NetParameter()

# load model
with open(proto_file, 'r') as f:
    text_format.Merge(f.read(), predict_net)
# load blob
with open(blob_file, 'rb') as f:
    init_net.ParseFromString(f.read())

shape_dict = {'data': [1,3,224,224]}
dtype_dict = {'data': 'float32'}

mod, params, model_outputs = relay.frontend.from_caffe(init_net, predict_net, shape_dict, dtype_dict)
```

### Work Done

All of the things that we have done are listed as following:

#### 1. List of supported Ops

- [x] BatchNorm
- [x] Concat
- [x] Convolution
- [x] Crop
- [x] Deconvolution
- [x] Dropout
- [x] Eltwise
- [x] Flatten
- [x] InnerProduct
- [x] Input
- [x] LRN
- [x] Normalize
- [x] Permute
- [x] Pooling
- [x] PReLU
- [x] PriorBox
- [x] proposal
- [x] Python
- [x] ReLU
- [x] Reshape
- [x] Resize
- [x] ROIPooling
- [x] Scale
- [x] Sigmoid
- [x] Slice
- [x] Softmax
- [x] TanH
- [x] Upsample

#### 2. List of supported complete models

- [x] Alexnet
- [x] Resnet50
- [x] Mobilenetv1
- [x] Mobilenetv2
- [x] Inceptionv1
- [x] Inceptionv3
- [x] Inceptionv4
- [x] Vgg16
- [x] Squeezenetv1
- [x] SSDMobilenetv1
- [x] SSDMobilenetv2
- [x] YOLOv3
- [x] ENet

### TODO

- [ ] Support more ops and more complete models.

According to the above implementation scheme, based on the front-end framework we built, you can add any new op, you only need to: firstly, add a method in the operatorconverter class, which needs to include your extraction of the layer parameters and the logic of conversion to TVM OP, secondly, register the method to convert_ map.

### CODE REVIEW

File: python/tvm/relay/frontend/caffe.py [hunk 1]
```diff
@@ -0,0 +1,1188 @@
...
+
+        scale = float(upsample_param.scale)
+
+        # get input expr
+        in_expr = self.exp_tab.get_expr(inputs[0])
+        # set params
+        scale_h_expr = scale
+        scale_w_expr = scale
+        # scale_h_expr = self.exp_tab.new_const(scale, dtype='float32')
+        # scale_w_expr = self.exp_tab.new_const(scale, dtype='float32')
```
Comment (FrozenGene) [inline python/tvm/relay/frontend/caffe.py:112]: remove

File: python/tvm/relay/frontend/caffe.py [hunk 2]
```diff
@@ -0,0 +1,1188 @@
...
+        params = dict()
+        params['layout'] = 'NCHW'
+        if len(inputs) == 1:
+            params['method'] = 'nearest_neighbor'
+            params['align_corners'] = False
+            out = _op.nn.upsampling(in_expr, scale_h_expr, scale_w_expr,
+                                    **params)
+        elif len(inputs) == 2:
+            # params['pad_out_h'] = 0
+            # params['pad_out_w'] = 0
```
Comment (FrozenGene) [inline python/tvm/relay/frontend/caffe.py:122]: remove

File: python/tvm/relay/frontend/caffe.py [hunk 3]
```diff
@@ -0,0 +1,1188 @@
...
+            output_tensors = pl.top
+
+            ret = self.convert_map[op_type](pl)
+
+            if len(output_tensors) == 1:
+                self.exp_tab.set_expr(output_tensors[0], ret)
+                logging.debug(
+                    "layer_name:{}, type:{}, output_name:{}, shape:{}".format(
+                        pl.name, pl.type, output_tensors[0],
+                        _infer_shape(ret)))
```
Comment (FrozenGene) [inline python/tvm/relay/frontend/caffe.py:1041]: don't add this debug information

File: python/tvm/relay/frontend/caffe.py [hunk 4]
```diff
@@ -0,0 +1,1188 @@
...
+    mod = IRModule.from_expr(func)
+    # return mod, params
+    new_outputs = list()
+    for i in model_outputs:
+        if i in top_change_before_dict:
+            tmp = top_change_before_dict[i]
+        else:
+            tmp = i
+        new_outputs.append(tmp)
+    return mod, params, new_outputs
```
Comment (FrozenGene) [inline python/tvm/relay/frontend/caffe.py:1188]: Please refer out other frontend code and we should only return `mod` and `params`.

**Comment by fernchen** [comment]
please refer to https://github.com/apache/incubator-tvm/pull/6206
